k9mail/src/main/java/com/fsck/k9/activity/KeywordEditor.java

```diff
@@ -0,0 +1,360 @@
+import android.widget.ImageButton;
+import android.widget.LinearLayout;
+import android.widget.ListView;
+import android.widget.TextView;
+import android.widget.Toast;
+import android.util.TypedValue;
+
+import com.fsck.k9.activity.ColorPickerDialog;
+import com.fsck.k9.fragment.KeywordAddDialogFragment;
+import com.fsck.k9.fragment.KeywordEditNameDialogFragment;
+import com.fsck.k9.K9;
+import com.fsck.k9.Preferences;
+import com.fsck.k9.R;
+import com.fsck.k9.mail.Keyword;
+
+
+public class KeywordEditor extends K9ListActivity implements
+    KeywordEditNameDialogFragment.KeywordEditNameDialogListener,
+    KeywordAddDialogFragment.KeywordAddDialogFragmentDialogListener
+{
+    private KeywordEditorListAdapter adapter;
+    private int colorKeywordNotVisible;
+
+
+    @Override
+    public void onCreate(Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+        setContentView(R.layout.keyword_editor_list);
+        ListView listView = getListView();
+        listView.setLongClickable(true);
+        registerForContextMenu(listView);
+    }
+
+    @Override
+    public boolean onCreateOptionsMenu(Menu menu) {
+        super.onCreateOptionsMenu(menu);
+        getMenuInflater().inflate(R.menu.keyword_editor_option, menu);
+        return true;
+    }
+
+    private void initializeActivityView() {
+
+        final ContextThemeWrapper themeContext = new ContextThemeWrapper(
+            this, K9.getK9ThemeResourceId(K9.getK9ComposerTheme()));
+        TypedValue outValue = new TypedValue();
+        themeContext.getTheme().resolveAttribute(
+            R.attr.colorKeywordNotVisible, outValue, true);
+        colorKeywordNotVisible = outValue.data;
+
+        ArrayList<Keyword> keywords = Keyword.getKeywords();
+        adapter = new KeywordEditorListAdapter(this, keywords);
+        setListAdapter(adapter);
+    }
+
+    @Override
+    public void onResume() {
+        super.onResume();
+
+        Preferences.getPreferences(this).loadKeywords();
+
+        if (adapter == null) {
+            initializeActivityView();
+        } else {
+            adapter.notifyDataSetChanged();
+        }
+    }
+
+    @Override
+    public void onBackPressed() {
+        setResult(RESULT_OK);
+        super.onBackPressed();
+    }
+
+    private void saveChangedKeyword(Keyword k) {
+        Preferences.getPreferences(this).saveKeyword(k);
+    }
+
+    private void saveChangedKeywordOrder() {
+        Preferences.getPreferences(this).saveKeywordOrder();
+    }
+
+    private void saveDeletedKeyword() {
+        Preferences.getPreferences(this).saveKeywords();
+    }
+
+    public class KeywordEditorListAdapter extends ArrayAdapter<Keyword> {
+
+        KeywordEditorListAdapter(Context context, List<Keyword> keywords) {
+            super(context, 0 /* resource not used */, keywords);
+        }
+
+        @Override
+        public View getView(
+            final int position, View convertView, ViewGroup parent)
+        {
```

Comment:
I agree. Now I refactored it and added a holder class to avoid repetition of one-time tasks. Still, the nested class KeywordEditorListAdapter has ~200 lines, which is almost half of the whole file. If desired, I can move the nested class into its own file.